src/Analyzers/Core/CodeFixes/NamespaceSync/AbstractNamespaceSyncCodeFixProvider.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Rename;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeFixes.NamespaceSync
+{
+    internal abstract partial class AbstractNamespaceSyncCodeFixProvider : CodeFixProvider
+    {
+        protected abstract string Title { get; }
+
+        public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(IDEDiagnosticIds.NamespaceSyncAnalyzerDiagnosticId);
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            context.RegisterCodeFix(
+                new MyCodeAction(Title, cancellationToken => FixAsync(context.Document, context.Diagnostics, cancellationToken)),
+                context.Diagnostics);
+
+            return Task.CompletedTask;
+        }
+
+        protected static async Task<Solution> FixAsync(Document document, ImmutableArray<Diagnostic> diagnostics, CancellationToken cancellationToken)
+        {
+            var currentDocument = document;
+
+            foreach (var diagnostic in diagnostics)
+            {
+                var targetNamespace = diagnostic.Properties["TargetNamespace"];
+                Contract.ThrowIfNull(targetNamespace);
+
+                var newSolution = await FixSingleDiagnosticAsync(currentDocument, targetNamespace, cancellationToken).ConfigureAwait(false);
+                currentDocument = newSolution.GetRequiredDocument(document.Id);
```

Comment:
i'm not sure how this is safe, since you're mutating the solution as you go and process each diagnostics.